handlers/ruleset.go

```diff
@@ -110,3 +144,125 @@ func ValidateRuleSet(store kv.Store, ruleSetKey string) (int, *schema.RuleSet, e
 	}
 	return ruleSetVersion, ruleSet, nil
 }
+
+// UpdateRules updates an existing RuleSet by appending a diffRuleSet to it.
+func UpdateRules(store kv.TxnStore,
+	ruleSet, diffRuleSet *schema.RuleSet,
+	ruleSetKey string, ruleSetVersion int,
+	namespacesKey string, namespacesVersion int,
+	propDelay time.Duration,
+) error {
+	// There is nothing to do if the diff is nil
+	if diffRuleSet == nil {
+		return nil
+	}
+
+	if ruleSet == nil {
+		ruleSet = diffRuleSet
+	}
+
+	err := appendToRuleSet(ruleSet, diffRuleSet)
+	if err != nil {
+		return err
+	}
+
+	nowNs := time.Now().UnixNano()
+	ruleSet.LastUpdatedAt = nowNs
+	ruleSet.CutoverTime = NewCutoverNs(nowNs, ruleSet.CutoverTime, propDelay)
+
+	// Perform a transaction and only update if the namespaces version
+	// and ruleSet version were unchanged.
+	namespacesCond := kv.NewCondition().
+		SetKey(namespacesKey).
+		SetCompareType(kv.CompareEqual).
+		SetTargetType(kv.TargetVersion).
+		SetValue(namespacesVersion)
+	ruleSetCond := kv.NewCondition().
+		SetKey(ruleSetKey).
+		SetCompareType(kv.CompareEqual).
+		SetTargetType(kv.TargetVersion).
+		SetValue(ruleSetVersion)
+	conditions := []kv.Condition{
+		namespacesCond,
+		ruleSetCond,
+	}
+	ops := []kv.Op{
+		kv.NewSetOp(ruleSetKey, ruleSet),
+	}
+
+	if _, err := store.Commit(conditions, ops); err != nil {
+		return fmt.Errorf("unable to update kv store: %v", err)
+	}
+
+	return nil
+}
+
+// AppendToRuleSet appends a ruleset diff onto the end of a ruleset
+func appendToRuleSet(orig, diff *schema.RuleSet) error {
+	for _, m := range diff.MappingRules {
+		mr, rr, err := RuleByID(orig, m.Uuid)
+		if err != nil && err != kv.ErrNotFound {
+			return err
+		}
+
+		if rr != nil {
+			return fmt.Errorf("rule with ID: %s is a rollup rule. Cannot make a mapping rule", rr.Uuid)
+		}
+
+		// A rule with no snapshots should not be added.
+		if m.Snapshots == nil {
+			continue
```

Comment:
This should be an error. Also if you are already checking the length below, you don't need this.